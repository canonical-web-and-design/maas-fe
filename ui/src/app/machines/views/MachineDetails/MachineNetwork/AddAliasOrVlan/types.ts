import type { NetworkValues } from "../NetworkFields/NetworkFields";

import type { NetworkInterface } from "app/store/machine/types";

export type AddAliasOrVlanValues = {
  tags?: NetworkInterface["tags"];
} & NetworkValues;
