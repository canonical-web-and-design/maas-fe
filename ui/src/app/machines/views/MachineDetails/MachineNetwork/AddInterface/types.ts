import type {
  NetworkInterface,
  NetworkLink,
  NetworkLinkMode,
  Vlan,
} from "app/store/machine/types";

export type AddInterfaceValues = {
  ip_address?: NetworkLink["ip_address"];
  mac_address: NetworkInterface["mac_address"];
  mode?: NetworkLinkMode;
  name?: NetworkInterface["name"];
  fabric: Vlan["fabric_id"];
  subnet?: NetworkLink["subnet_id"];
  tags?: NetworkInterface["tags"];
  vlan: NetworkInterface["vlan_id"];
};
