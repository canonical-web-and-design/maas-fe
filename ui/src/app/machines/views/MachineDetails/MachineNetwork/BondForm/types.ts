import type { NetworkValues } from "../NetworkFields/NetworkFields";

import type {
  BondLacpRate,
  BondMode,
  BondXmitHashPolicy,
} from "app/store/general/types";
import type {
  NetworkInterface,
  NetworkInterfaceParams,
} from "app/store/machine/types";

export enum MIIOptions {
  MII = "mii",
}

export type BondFormValues = {
  bond_downdelay: NetworkInterfaceParams["bond_downdelay"];
  bond_lacp_rate: BondLacpRate;
  bond_miimon: NetworkInterfaceParams["bond_miimon"];
  bond_mode: BondMode;
  bond_updelay: NetworkInterfaceParams["bond_updelay"];
  bond_xmit_hash_policy: BondXmitHashPolicy;
  linkMonitoring: string;
  mac_address: NetworkInterface["mac_address"];
  name: NetworkInterface["name"];
  primary: NetworkInterface["id"];
  tags: NetworkInterface["tags"];
} & NetworkValues;
