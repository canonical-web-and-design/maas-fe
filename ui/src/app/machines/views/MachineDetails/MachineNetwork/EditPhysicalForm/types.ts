import type { NetworkValues } from "../NetworkFields/NetworkFields";

import type { NetworkInterface } from "app/store/machine/types";

export type EditPhysicalValues = {
  interface_speed: NetworkInterface["interface_speed"];
  link_speed: NetworkInterface["link_speed"];
  mac_address: NetworkInterface["mac_address"];
  name?: NetworkInterface["name"];
  tags?: NetworkInterface["tags"];
} & NetworkValues;
