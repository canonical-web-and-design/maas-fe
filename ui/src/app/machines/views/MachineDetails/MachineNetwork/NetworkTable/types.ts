import type { NetworkInterface, NetworkLink } from "app/store/types/node";

export enum ExpandedState {
  ADD_ALIAS = "addAlias",
  ADD_BOND = "addBond",
  ADD_BRIDGE = "addBridge",
  ADD_PHYSICAL = "addPhysical",
  ADD_VLAN = "addVlan",
  DISCONNECTED_WARNING = "disconnectedWarning",
  EDIT = "edit",
  MARK_CONNECTED = "markConnected",
  MARK_DISCONNECTED = "markDisconnected",
  REMOVE = "remove",
}

export type Expanded = {
  content: ExpandedState;
  linkId?: NetworkLink["id"] | null;
  nicId?: NetworkInterface["id"] | null;
};

export type SetExpanded = (expanded: Expanded | null) => void;

export type Selected = {
  linkId?: NetworkLink["id"] | null;
  nicId?: NetworkInterface["id"] | null;
};

export type SetSelected = (selected: Selected[]) => void;
