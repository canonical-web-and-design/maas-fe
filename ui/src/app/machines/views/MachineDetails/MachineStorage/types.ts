import type { Filesystem } from "app/store/machine/types";

export type NormalisedFilesystem = {
  fstype: Filesystem["fstype"];
  id: Filesystem["id"];
  mountOptions: Filesystem["mount_options"];
  mountPoint: Filesystem["mount_point"];
  name: string | null;
  size: number | null;
};

export type NormalisedStorageDevice = {
  boot: boolean | null;
  firmware: string | null;
  id: number;
  model: string | null;
  name: string;
  numaNodes: number[];
  parentType: string | null;
  serial: string | null;
  size: number;
  tags: string[];
  testStatus: number | null;
  type: string;
  usedFor: string;
};

export type SeparatedDiskData = {
  available: NormalisedStorageDevice[];
  cacheSets: NormalisedStorageDevice[];
  datastores: NormalisedFilesystem[];
  filesystems: NormalisedFilesystem[];
  used: NormalisedStorageDevice[];
};
