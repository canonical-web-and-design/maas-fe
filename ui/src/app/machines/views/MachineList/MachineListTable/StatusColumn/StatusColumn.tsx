import * as React from "react";

import { Spinner, Tooltip } from "@canonical/react-components";
import PropTypes from "prop-types";
import { useSelector } from "react-redux";
import { Link } from "react-router-dom";

import DoubleRow from "app/base/components/DoubleRow";
import { useMachineActions } from "app/base/hooks";
import { useToggleMenu } from "app/machines/hooks";
import machineSelectors from "app/store/machine/selectors";
import type { Machine } from "app/store/machine/types";
import { isTransientStatus, useFormattedOS } from "app/store/machine/utils";
import type { RootState } from "app/store/root/types";
import {
  NodeActions,
  NodeStatusCode,
  TestStatusStatus,
} from "app/store/types/node";
import { getStatusText } from "app/utils";

// Node statuses for which the failed test warning is not shown.
const hideFailedTestWarningStatuses = [
  NodeStatusCode.COMMISSIONING,
  NodeStatusCode.FAILED_COMMISSIONING,
  NodeStatusCode.FAILED_TESTING,
  NodeStatusCode.NEW,
  NodeStatusCode.TESTING,
];

const getProgressText = (machine: Machine) => {
  if (isTransientStatus(machine.status_code)) {
    return machine.status_message;
  }
  return "";
};

const getStatusIcon = (machine: Machine) => {
  if (isTransientStatus(machine.status_code)) {
    return <Spinner data-test="status-icon" />;
  } else if (
    machine.testing_status.status === TestStatusStatus.FAILED &&
    !hideFailedTestWarningStatuses.includes(machine.status_code)
  ) {
    return (
      <Tooltip
        message="Machine has failed tests; use with caution."
        position="top-left"
      >
        <i className="p-icon--warning" data-test="status-icon" />
      </Tooltip>
    );
  }
  return "";
};

type Props = {
  onToggleMenu?: (systemId: string, open: boolean) => void;
  systemId: string;
};

export const StatusColumn = ({
  onToggleMenu,
  systemId,
}: Props): JSX.Element | null => {
  const machine = useSelector((state: RootState) =>
    machineSelectors.getById(state, systemId)
  );
  const formattedOS = useFormattedOS(machine);
  const toggleMenu = useToggleMenu(onToggleMenu, systemId);
  const actionLinks = useMachineActions(systemId, [
    NodeActions.ABORT,
    NodeActions.ACQUIRE,
    NodeActions.COMMISSION,
    NodeActions.DEPLOY,
    NodeActions.EXIT_RESCUE_MODE,
    NodeActions.LOCK,
    NodeActions.MARK_BROKEN,
    NodeActions.MARK_FIXED,
    NodeActions.OVERRIDE_FAILED_TESTING,
    NodeActions.RELEASE,
    NodeActions.RESCUE_MODE,
    NodeActions.TEST,
    NodeActions.UNLOCK,
  ]);

  const statusText = getStatusText(machine, formattedOS);
  const menuLinks = [
    actionLinks,
    [
      {
        children: "See logs",
        element: Link,
        to: `/machine/${systemId}/logs`,
      },
    ],
  ];

  if (machine) {
    return (
      <DoubleRow
        icon={getStatusIcon(machine)}
        iconSpace={true}
        menuLinks={onToggleMenu && menuLinks}
        menuTitle="Take action:"
        onToggleMenu={toggleMenu}
        primary={
          <span data-test="status-text" title={statusText}>
            {statusText}
          </span>
        }
        secondary={
          <>
            <span data-test="progress-text" title={getProgressText(machine)}>
              {getProgressText(machine)}
            </span>
            <span data-test="error-text">
              {machine.error_description &&
              machine.status_code === NodeStatusCode.BROKEN
                ? machine.error_description
                : ""}
            </span>
          </>
        }
      />
    );
  }
  return null;
};

StatusColumn.propTypes = {
  onToggleMenu: PropTypes.func,
  systemId: PropTypes.string.isRequired,
};

export default React.memo(StatusColumn);
