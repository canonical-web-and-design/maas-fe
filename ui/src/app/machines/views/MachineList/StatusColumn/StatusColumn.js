import { Loader } from "@canonical/react-components";
import { useSelector } from "react-redux";
import React from "react";
import PropTypes from "prop-types";

import {
  general as generalSelectors,
  machine as machineSelectors
} from "app/base/selectors";
import { nodeStatus, scriptStatus } from "app/base/enum";
import Tooltip from "app/base/components/Tooltip";

// Node statuses for which the failed test warning is not shown.
const hideFailedTestWarningStatuses = [
  nodeStatus.COMMISSIONING,
  nodeStatus.FAILED_COMMISSIONING,
  nodeStatus.FAILED_TESTING,
  nodeStatus.NEW,
  nodeStatus.TESTING
];

// Node statuses for which the OS + release is made human-readable.
const formattedReleaseStatuses = [nodeStatus.DEPLOYED, nodeStatus.DEPLOYING];

// Node statuses that are temporary.
const transientStatuses = [
  nodeStatus.COMMISSIONING,
  nodeStatus.DEPLOYING,
  nodeStatus.DISK_ERASING,
  nodeStatus.ENTERING_RESCUE_MODE,
  nodeStatus.EXITING_RESCUE_MODE,
  nodeStatus.RELEASING,
  nodeStatus.TESTING
];

// Script statuses associated with failure.
const failedScriptStatuses = [
  scriptStatus.DEGRADED,
  scriptStatus.FAILED,
  scriptStatus.FAILED_APPLYING_NETCONF,
  scriptStatus.FAILED_INSTALLING,
  scriptStatus.TIMEDOUT
];

const getStatusText = (machine, osReleases) => {
  if (formattedReleaseStatuses.includes(machine.status_code)) {
    const machineRelease = osReleases.find(
      release => release.value === machine.distro_series
    );

    if (machineRelease) {
      let releaseTitle;
      if (machine.osystem === "ubuntu") {
        releaseTitle = machineRelease.label.split('"')[0].trim();
      } else {
        releaseTitle = machineRelease.label;
      }

      if (machine.status_code === nodeStatus.DEPLOYING) {
        return `Deploying ${releaseTitle}`;
      }
      return releaseTitle;
    }
  }
  return machine.status;
};

const getProgressText = machine => {
  if (transientStatuses.includes(machine.status_code)) {
    return machine.status_message;
  }
  return "";
};

const getStatusIcon = machine => {
  if (transientStatuses.includes(machine.status_code)) {
    return (
      <Loader
        className="u-no-margin u-no-padding"
        data-test="status-icon"
        inline
      />
    );
  } else if (
    failedScriptStatuses.includes(machine.testing_status.status) &&
    !hideFailedTestWarningStatuses.includes(machine.status_code)
  ) {
    return (
      <Tooltip
        message="Machine has failed tests; use with caution."
        position="top-left"
      >
        <i className="p-icon--warning" data-test="status-icon" />
      </Tooltip>
    );
  }
  return "";
};

const StatusColumn = ({ systemId }) => {
  const machine = useSelector(state =>
    machineSelectors.getBySystemId(state, systemId)
  );
  const osReleases = useSelector(state =>
    generalSelectors.osInfo.getOsReleases(state, machine.osystem)
  );

  return (
    <div className="p-double-row--with-icon">
      <div className="p-double-row__icon">{getStatusIcon(machine)}</div>
      <div
        className="p-double-row__primary-row u-truncate"
        data-test="status-text"
        title={getStatusText(machine, osReleases)}
      >
        {getStatusText(machine, osReleases)}
      </div>
      <div
        className="p-double-row__secondary-row u-truncate"
        data-test="progress-text"
        title={getProgressText(machine)}
      >
        {getProgressText(machine)}
      </div>
    </div>
  );
};

StatusColumn.propTypes = {
  systemId: PropTypes.string.isRequired
};

export default StatusColumn;
