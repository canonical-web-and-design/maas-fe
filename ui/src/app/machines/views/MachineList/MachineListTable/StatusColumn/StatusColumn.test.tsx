import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import React from "react";

import { nodeStatus, scriptStatus } from "app/base/enum";
import { StatusColumn } from "./StatusColumn";
import type { Machine } from "app/store/machine/types";
import type { RootState } from "app/store/root/types";
import type { TestResult } from "app/store/types/node";
import { NodeStatus } from "app/store/types/node";
import {
  machine as machineFactory,
  machineState as machineStateFactory,
  generalState as generalStateFactory,
  osInfo as osInfoFactory,
  osInfoState as osInfoStateFactory,
  rootState as rootStateFactory,
} from "testing/factories";

const mockStore = configureStore();

describe("StatusColumn", () => {
  let state: RootState;
  let machine: Machine;

  beforeEach(() => {
    machine = machineFactory({
      actions: [],
      distro_series: "bionic",
      osystem: "ubuntu",
      status: NodeStatus.NEW,
      status_code: 0,
      status_message: "",
      system_id: "abc123",
    });
    state = rootStateFactory({
      general: generalStateFactory({
        osInfo: osInfoStateFactory({
          data: osInfoFactory({
            osystems: [
              ["centos", "CentOS"],
              ["ubuntu", "Ubuntu"],
            ],
            releases: [
              ["centos/centos70", "CentOS 7"],
              ["ubuntu/xenial", 'Ubuntu 16.04 LTS "Xenial Xerus"'],
              ["ubuntu/bionic", 'Ubuntu 18.04 LTS "Bionic Beaver"'],
            ],
          }),
          errors: {},
          loaded: true,
          loading: false,
        }),
      }),
      machine: machineStateFactory({
        items: [machine],
        errors: {},
        loaded: true,
        loading: false,
      }),
    });
  });

  it("renders", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("StatusColumn")).toMatchSnapshot();
  });

  describe("status text", () => {
    it("displays the machine's status if not deploying or deployed", () => {
      machine.status = NodeStatus.NEW;
      machine.status_code = nodeStatus.NEW;
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find("[data-test='status-text']").text()).toBe("New");
    });

    it("displays the short-form of Ubuntu release if deployed", () => {
      machine.status = NodeStatus.DEPLOYED;
      machine.status_code = nodeStatus.DEPLOYED;
      machine.osystem = "ubuntu";
      machine.distro_series = "bionic";
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find("[data-test='status-text']").text()).toBe(
        "Ubuntu 18.04 LTS"
      );
    });

    it("displays the full OS and release if non-Ubuntu deployed", () => {
      machine.status = NodeStatus.DEPLOYED;
      machine.status_code = nodeStatus.DEPLOYED;
      machine.osystem = "centos";
      machine.distro_series = "centos70";
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find("[data-test='status-text']").text()).toBe("CentOS 7");
    });

    it("displays 'Deploying OS release' if machine is deploying", () => {
      machine.status = NodeStatus.DEPLOYING;
      machine.status_code = nodeStatus.DEPLOYING;
      machine.osystem = "ubuntu";
      machine.distro_series = "bionic";
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find("[data-test='status-text']").text()).toBe(
        "Deploying Ubuntu 18.04 LTS"
      );
    });

    it("displays an error message for broken machines", () => {
      machine.error_description = "machine is on fire";
      machine.status = NodeStatus.BROKEN;
      machine.status_code = nodeStatus.BROKEN;
      const store = mockStore(state);

      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find("[data-test='error-text']").text()).toBe(
        "machine is on fire"
      );
    });
  });

  describe("progress text", () => {
    it("displays the machine's status_message if in a transient state", () => {
      machine.status = NodeStatus.TESTING;
      machine.status_code = nodeStatus.TESTING;
      machine.status_message = "2 of 6 tests complete";
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find("[data-test='progress-text']").text()).toBe(
        "2 of 6 tests complete"
      );
    });

    it(`does not display the machine's status_message if
      not in a transient state`, () => {
      machine.status = NodeStatus.ALLOCATED;
      machine.status_code = nodeStatus.ALLOCATED;
      machine.status_message = "This machine is allocated";
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find("[data-test='progress-text']").text()).toBe("");
    });
  });

  describe("status icon", () => {
    it("shows a spinner if machine is in a transient state", () => {
      machine.status = NodeStatus.COMMISSIONING;
      machine.status_code = nodeStatus.COMMISSIONING;
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find(".p-icon--spinner").exists()).toBe(true);
    });

    it(`shows a warning and tooltip if machine has failed tests and is not in a
      state where the warning should be hidden`, () => {
      machine.status = NodeStatus.ALLOCATED;
      machine.status_code = nodeStatus.ALLOCATED;
      machine.testing_status.status = scriptStatus.FAILED as TestResult;
      const store = mockStore(state);
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter
            initialEntries={[{ pathname: "/machines", key: "testKey" }]}
          >
            <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find(".p-icon--warning").exists()).toBe(true);
      expect(wrapper.find("Tooltip").exists()).toBe(true);
    });
  });

  it("can show a menu with all possible options", () => {
    machine.actions = [
      "abort",
      "acquire",
      "commission",
      "deploy",
      "exit-rescue-mode",
      "lock",
      "mark-broken",
      "mark-fixed",
      "override-failed-testing",
      "release",
      "rescue-mode",
      "test",
      "unlock",
    ];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <StatusColumn onToggleMenu={jest.fn()} systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );
    // Open the menu so the elements get rendered.
    wrapper.find("Button.p-contextual-menu__toggle").simulate("click");
    expect(wrapper.find(".p-contextual-menu__dropdown")).toMatchSnapshot();
  });

  it("does not render table menu if onToggleMenu not provided", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <StatusColumn systemId="abc123" />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find("TableMenu").exists()).toBe(false);
  });
});
