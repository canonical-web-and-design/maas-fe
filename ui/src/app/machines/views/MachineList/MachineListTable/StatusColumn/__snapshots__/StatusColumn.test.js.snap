// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusColumn can show a menu with all possible options 1`] = `
<span
  aria-hidden="false"
  aria-label="submenu"
  className="p-contextual-menu__dropdown"
  id="Uakgb_J5m9g-0JDMbcJqLJ"
>
  <div
    className="p-contextual-menu__non-interactive"
    key="0"
  >
    Take action:
  </div>
  <span
    className="p-contextual-menu__group"
    key="1"
  >
    <Button
      className="p-contextual-menu__link"
      key="0"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        abort
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="1"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        acquire
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="2"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        commission
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="3"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        deploy
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="4"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        exit-rescue-mode
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="5"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        lock
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="6"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        mark-broken
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="7"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        mark-fixed
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="8"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        override-failed-testing
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="9"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        release
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="10"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        rescue-mode
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="11"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        test
      </button>
    </Button>
    <Button
      className="p-contextual-menu__link"
      key="12"
      onClick={[Function]}
    >
      <button
        className="p-button--neutral p-contextual-menu__link"
        onClick={[Function]}
      >
        unlock
      </button>
    </Button>
  </span>
  <span
    className="p-contextual-menu__group"
    key="2"
  >
    <Button
      className="p-contextual-menu__link"
      element="a"
      href="/MAAS/l/machine/abc123?area=logs"
      key="0"
      onClick={null}
    >
      <a
        className="p-button--neutral p-contextual-menu__link"
        href="/MAAS/l/machine/abc123?area=logs"
        onClick={null}
      >
        See logs
      </a>
    </Button>
    <Button
      className="p-contextual-menu__link"
      element="a"
      href="/MAAS/l/machine/abc123?area=events"
      key="1"
      onClick={null}
    >
      <a
        className="p-button--neutral p-contextual-menu__link"
        href="/MAAS/l/machine/abc123?area=events"
        onClick={null}
      >
        See events
      </a>
    </Button>
  </span>
</span>
`;

exports[`StatusColumn renders 1`] = `
<StatusColumn
  onToggleMenu={[MockFunction]}
  systemId="abc123"
>
  <DoubleRow
    icon=""
    iconSpace={true}
    menuLinks={
      Array [
        Array [],
        Array [
          Object {
            "children": "See logs",
            "element": "a",
            "href": "/MAAS/l/machine/abc123?area=logs",
          },
          Object {
            "children": "See events",
            "element": "a",
            "href": "/MAAS/l/machine/abc123?area=events",
          },
        ],
      ]
    }
    menuTitle="Take action:"
    onToggleMenu={[Function]}
    primary={
      <span
        data-test="status-text"
        title="New"
      >
        New
      </span>
    }
    secondary={
      <span
        data-test="progress-text"
        title=""
      >
        
      </span>
    }
  >
    <div
      className="p-double-row--with-icon"
    >
      <div
        className="p-double-row__icon"
      >
        <div
          className="p-double-row__icon-space"
        />
      </div>
      <div
        className="p-double-row__rows-container"
      >
        <div
          className="p-double-row__primary-row"
        >
          <div
            className="p-double-row__primary-row-text u-truncate"
          >
            <span
              data-test="status-text"
              title="New"
            >
              New
            </span>
          </div>
          <TableMenu
            links={
              Array [
                Array [],
                Array [
                  Object {
                    "children": "See logs",
                    "element": "a",
                    "href": "/MAAS/l/machine/abc123?area=logs",
                  },
                  Object {
                    "children": "See events",
                    "element": "a",
                    "href": "/MAAS/l/machine/abc123?area=events",
                  },
                ],
              ]
            }
            onToggleMenu={[Function]}
            positionNode={
              Object {
                "current": <div
                  class="p-double-row__primary-row"
                >
                  <div
                    class="p-double-row__primary-row-text u-truncate"
                  >
                    <span
                      data-test="status-text"
                      title="New"
                    >
                      New
                    </span>
                  </div>
                  <span
                    class="p-table-menu p-contextual-menu--left"
                  >
                    <button
                      aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                      aria-expanded="false"
                      aria-haspopup="true"
                      class="p-button--base has-icon p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                      type="button"
                    >
                      <i
                        class="p-icon--contextual-menu p-contextual-menu__indicator"
                      />
                    </button>
                  </span>
                </div>,
              }
            }
            title="Take action:"
          >
            <ContextualMenu
              className="p-table-menu"
              hasToggleIcon={true}
              links={
                Array [
                  "Take action:",
                  Array [],
                  Array [
                    Object {
                      "children": "See logs",
                      "element": "a",
                      "href": "/MAAS/l/machine/abc123?area=logs",
                    },
                    Object {
                      "children": "See events",
                      "element": "a",
                      "href": "/MAAS/l/machine/abc123?area=events",
                    },
                  ],
                ]
              }
              onToggleMenu={[Function]}
              position="left"
              positionNode={
                Object {
                  "current": <div
                    class="p-double-row__primary-row"
                  >
                    <div
                      class="p-double-row__primary-row-text u-truncate"
                    >
                      <span
                        data-test="status-text"
                        title="New"
                      >
                        New
                      </span>
                    </div>
                    <span
                      class="p-table-menu p-contextual-menu--left"
                    >
                      <button
                        aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                        aria-expanded="false"
                        aria-haspopup="true"
                        class="p-button--base has-icon p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                        type="button"
                      >
                        <i
                          class="p-icon--contextual-menu p-contextual-menu__indicator"
                        />
                      </button>
                    </span>
                  </div>,
                }
              }
              toggleAppearance="base"
              toggleClassName="u-no-margin--bottom p-table-menu__toggle"
            >
              <span
                className="p-table-menu p-contextual-menu--left"
              >
                <Button
                  appearance="base"
                  aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                  aria-expanded="false"
                  aria-haspopup="true"
                  className="p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                  hasIcon={true}
                  onClick={[Function]}
                  type="button"
                >
                  <button
                    aria-controls="Uakgb_J5m9g-0JDMbcJqLJ"
                    aria-expanded="false"
                    aria-haspopup="true"
                    className="p-button--base has-icon p-contextual-menu__toggle u-no-margin--bottom p-table-menu__toggle"
                    onClick={[Function]}
                    type="button"
                  >
                    <i
                      className="p-icon--contextual-menu p-contextual-menu__indicator"
                    />
                  </button>
                </Button>
              </span>
            </ContextualMenu>
          </TableMenu>
        </div>
        <div
          className="p-double-row__secondary-row u-truncate"
        >
          <span
            data-test="progress-text"
            title=""
          />
        </div>
      </div>
    </div>
  </DoubleRow>
</StatusColumn>
`;
